MPI

mahdiqezlou opened the pull request:
The code, is still raw and not very generalized; for instance, the hdf5 files are spread between different ranks manually by grouping them in different folders (I will generalize it later, so we can focus just on the parallel part now). 

Something is specially odd : When I run the code with sufficient cores (19), rank = 0 never runs. I have had some other MPI check tests in which all ranks have run very well.

To run the program, I just run "parallel.py" file with .sh file below :

#!/bin/bash

#SBATCH --partition=short
#SBATCH --nodes=1
#SBATCH --tasks-per-node=19
#SBATCH --mem-per-cpu=3G
#SBATCH --cpus-per-task=1
#SBATCH -p short
#SBATCH --job-name="parallel"

#SBATCH --time=01:00:00

module load mpich

mpirun python parallel.py

sbird commented on fake_spectra/abstractsnapshot.py:
```diff
@@ -164,10 +164,11 @@ def _get_all_files(self, num, base):
         snap=str(num).rjust(3,'0')
-        new_fname = os.path.join(base, "snapdir_"+snap)
+        #new_fname = os.path.join(base, "snapdir_"+snap)
+        new_fname = base
```
This will break existing scripts. Could you instead fix loading snapshots called something else by adding an extra argument (call it snapprefix) to the constructor of AbstractSnapshot and propagating it down? HDF5Snapshot can have a default set as snapprefix = "snapdir_" and BigFileSnapshot can have as a default snapprefix="PART_"

Then _get_all_files can do:
new_fname = os.path.join(base, snapprefix+snap)

sbird commented on fake_spectra/spectra.py:
```diff
@@ -606,11 +612,21 @@ def replace_not_DLA(self, ndla, thresh=10**20.3, elem="H", ion=1):
+        self.comm.Reduce(cdsum, col_den_added, op=self.MPI.SUM, root=0)
+        size_ind = self.comm.bcast(size_ind, root=0)
+        ind = np.zeros(size_ind, dtype='d')
+        self.comm.Bcast(ind, root=0)
```
You are broadcasting an array which is zero here. Is this placeholder code for future reimplementation of the filter_DLA function?

sbird commented on fake_spectra/spectra.py:
```diff
@@ -619,18 +635,32 @@ def replace_not_DLA(self, ndla, thresh=10**20.3, elem="H", ion=1):
+            #ind = self.filter_DLA(col_den, thresh)
+            
+            ### Call manager rank here
+            self.comm.Reduce(cdsum, col_den_added, op=MPI.SUM, root=0)
```
You could put this and the above behind a condition:
if self.MPI:
(self.comm.Reduce...)

sbird commented:
So you are saying that the if rank == 0: condition in parallel.py is always false if and only if size > 16? That is extremely strange, not to mention awkward.

sbird commented on fake_spectra/parallel.py:
```diff
@@ -0,0 +1,85 @@
+        col_den_added = np.zeros(shape=(numlos,))
+
+        ### Recieve col_den from all other ranks and add them together
+        comm.Reduce(np.zeros(shape=(numlos,), dtype='d'), col_den_added, op = MPI.SUM, root = 0)
```
This function needs to be called on all ranks to be effective. When the function is called, what happens is something like this:
if rank == root:
MPI_Receive(data, all_ranks)
sum(data)
else:
MPI_Send(data, root)

Because this is not a shared memory system the data cannot be simply read by the master rank: it has to be sent from the other ranks as well. A good rule for MPI is that all collective operations have to be called by every rank in the communicator.

This is because otherwise the code doesn't know at what point in execution to read data: these processes are not guaranteed to run in any particular order. So the Send call means "get here and then send the message".

mahdiqezlou replied:
> This function needs to be called on all ranks to be effective. When the function is called, what happens is something like this:
> if rank == root:
> MPI_Receive(data, all_ranks)
> sum(data)
> else:
> MPI_Send(data, root)
> 
> Because this is not a shared memory system the data cannot be simply read by the master rank: it has to be sent from the other ranks as well. A good rule for MPI is that all collective operations have to be called by every rank in the communicator.
> 
> This is because otherwise the code doesn't know at what point in execution to read data: these processes are not guaranteed to run in any particular order. So the Send call means "get here and then send the message".

In replace_not_DLA  in spectra.py , I have called comm.Reduce(colden, colden_added, ...). So, after calculating the column densities in each rank they sent back the column densities to rank = 0. Then rank =0 returns all indices which are above threshold to all labor ranks, then they are able to replace those ranks.
So, do you mean the comm.Reduce() should be called in other places than replace_not_DLA ?

sbird commented on fake_spectra/parallel.py:
```diff
@@ -0,0 +1,85 @@
+
+
+        size_ind = comm.bcast(size_ind, root=0)
+        comm.Bcast(ind, root = 0)
```
This broadcast too: it needs to be called on the other ranks so that the other ranks know when to receive the data.

mahdiqezlou replied:
Yes, I see your point. As I explained about comm.Replace() in above comment, I have called comm.Bcast(ind , root = 0) in replace_not_DLA().
For labor ranks, would there be any difference if I call comm.Bcast(ind, root= 0)  in parallel.py instead of replace_not_DLA() ?

mahdiqezlou commented:
> I think I see the problem: you have a master/slave architecture and you have the routines to do the receive, but not yet the routines to do the send. So a reorganisation of parallel.py could look like:
> 
> #On all ranks
> snap = RandSpectra(...)
> tt = snap.get_tau(...)
> #(all the code currently inside if rank != 0)
> col_den_added = np.zeros(...)
> #call reduce on all ranks.
> comm.Reduce(np.zeros(shape=(numlos,), dtype='d'), col_den_added, op = MPI.SUM, root = 0)
> if rank ==0:
> (calculate found)
> #Call broadcast function on all ranks.
> comm.Bcast(ind, root=0)

Yes, rank = 0 just adds column densities found by other ranks. So, by this reorganization, you mean even rank = 0 should do the column density calculations besides adding them, am I right ?

sbird commented:
Ok, I understand now. The reduce() in rank == 0 is matched to a reduce() in the other ranks in replace_not_DLA(). One thing that occurred to me on the way over here is that some MPI libraries have a requirement that the calls have to happen at roughly similar times. So if the rank == 0 part of the reduce happens much later than the rank != 0 part, the call might have timed out. If that is the case, computing the column density calculations on rank 0 as well might fix it.

mahdiqezlou commented:
> Ok, I understand now. The reduce() in rank == 0 is matched to a reduce() in the other ranks in replace_not_DLA(). One thing that occurred to me on the way over here is that some MPI libraries have a requirement that the calls have to happen at roughly similar times. So if the rank == 0 part of the reduce happens much later than the rank != 0 part, the call might have timed out. If that is the case, computing the column density calculations on rank 0 as well might fix it.

Oh, I see. Sure, I'll let rank = 0 do column density calculations, as well.

sbird commented:
I just ran this branch on a small box on my workstation (with the print statements added) and it completed fine. So the code itself does not look to be buggy.

sbird commented:
Try adding "del rr" as the last line of parallel.py : this will get rid of the error on exit.

mahdiqezlou commented:
Oh Nice.Thanks!
 Could you please share your submission file ( omp_submit.sh) ?
So, it suggests that problem is cause by running on cluster.
I am going to run on small snapshots on cluster as well.

On Tue, Jul 2, 2019, 3:12 PM Simeon Bird <notifications@github.com> wrote:

> I just ran this branch on a small box on my workstation (with the print
> statements added) and it completed fine. So the code itself does not look
> to be buggy.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sbird/fake_spectra/pull/22?email_source=notifications&email_token=AKSFRZ35G2VLPTWTR27AA7LP5PHDVA5CNFSM4H4U6XE2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZCWHLA#issuecomment-507863980>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKSFRZ6VMRQFFHZ6P5SC7VLP5PHDVANCNFSM4H4U6XEQ>
> .
>

sbird commented:
I was just running it on my workstation so there is no submission to do: I just type 

OMP_NUM_THREADS=2 mpirun -np 4 python parallel.py
and wait

sbird commented:
There is a problem with thresh > 0 though: if I set thresh=thresh_t I get this backtrace:

Traceback (most recent call last):
  File "parallel.py", line 20, in <module>
    rr = RandSpectra(5, "/home/spb/data/Cosmo/Cosmo7_V6/L25n512/output/snapdir_005/", MPI, comm,  thresh = thresh_t, kernel='tophot',ndla = ndla, numlos=numlos,savedir="/home/spb/data/Cosmo/Cosmo7_V6/L25n512/output/parallel", savefile="spectra_34."+rank_str+".hdf5")
  File "/home/spb/.local/lib/python3.7/site-packages/fake_spectra/randspectra.py", line 30, in __init__
    self.replace_not_DLA(ndla, thresh, elem=elem, ion=ion)
  File "/home/spb/.local/lib/python3.7/site-packages/fake_spectra/spectra.py", line 642, in replace_not_DLA
    cofm_DLA[found:top] = self.cofm[ind][:top,:]
IndexError: arrays used as indices must be of integer (or boolean) type
Traceback (most recent call last):
  File "parallel.py", line 20, in <module>
    rr = RandSpectra(5, "/home/spb/data/Cosmo/Cosmo7_V6/L25n512/output/snapdir_005/", MPI, comm,  thresh = thresh_t, kernel='tophot',ndla = ndla, numlos=numlos,savedir="/home/spb/data/Cosmo/Cosmo7_V6/L25n512/output/parallel", savefile="spectra_34."+rank_str+".hdf5")
  File "/home/spb/.local/lib/python3.7/site-packages/fake_spectra/randspectra.py", line 30, in __init__
    self.replace_not_DLA(ndla, thresh, elem=elem, ion=ion)
  File "/home/spb/.local/lib/python3.7/site-packages/fake_spectra/spectra.py", line 642, in replace_not_DLA
    cofm_DLA[found:top] = self.cofm[ind][:top,:]
IndexError: arrays used as indices must be of integer (or boolean) type
Traceback (most recent call last):
  File "parallel.py", line 20, in <module>
    rr = RandSpectra(5, "/home/spb/data/Cosmo/Cosmo7_V6/L25n512/output/snapdir_005/", MPI, comm,  thresh = thresh_t, kernel='tophot',ndla = ndla, numlos=numlos,savedir="/home/spb/data/Cosmo/Cosmo7_V6/L25n512/output/parallel", savefile="spectra_34."+rank_str+".hdf5")
  File "/home/spb/.local/lib/python3.7/site-packages/fake_spectra/randspectra.py", line 30, in __init__
    self.replace_not_DLA(ndla, thresh, elem=elem, ion=ion)
  File "/home/spb/.local/lib/python3.7/site-packages/fake_spectra/spectra.py", line 642, in replace_not_DLA
    cofm_DLA[found:top] = self.cofm[ind][:top,:]
IndexError: arrays used as indices must be of integer (or boolean) type
Traceback (most recent call last):
  File "parallel.py", line 20, in <module>
    rr = RandSpectra(5, "/home/spb/data/Cosmo/Cosmo7_V6/L25n512/output/snapdir_005/", MPI, comm,  thresh = thresh_t, kernel='tophot',ndla = ndla, numlos=numlos,savedir="/home/spb/data/Cosmo/Cosmo7_V6/L25n512/output/parallel", savefile="spectra_34."+rank_str+".hdf5")
  File "/home/spb/.local/lib/python3.7/site-packages/fake_spectra/randspectra.py", line 30, in __init__
    self.replace_not_DLA(ndla, thresh, elem=elem, ion=ion)
  File "/home/spb/.local/lib/python3.7/site-packages/fake_spectra/spectra.py", line 638, in replace_not_DLA
    self.comm.Bcast(ind, root=0)
  File "mpi4py/MPI/Comm.pyx", line 578, in mpi4py.MPI.Comm.Bcast
  File "mpi4py/MPI/msgbuffer.pxi", line 472, in mpi4py.MPI._p_msg_cco.for_bcast
  File "mpi4py/MPI/msgbuffer.pxi", line 436, in mpi4py.MPI._p_msg_cco.for_cco_send
  File "mpi4py/MPI/msgbuffer.pxi", line 145, in mpi4py.MPI.message_simple
ValueError: message: expecting 2 to 4 items

sbird commented:
which I think is because of this in spectra.py:
if self.rank != 0 : 
ind = np.zeros(size_ind, dtype='d')

sbird commented:
You want self.rank instead of rank in a few places too

sbird commented:
The self.comm.Bcast() call is complicated, because ind is a 1-length tuple: (np.array([0,1,2,...]),) so changing the type is not enough

sbird commented:
I think you want code like:

        if self.rank != 0 :
            ind = np.zeros(size_ind, dtype=np.int)
        else:
            ind = ind[0]

mahdiqezlou commented:
> I think you want code like:
> 
> ```
>     if self.rank != 0 :
>         ind = np.zeros(size_ind, dtype=np.int)
>     else:
>         ind = ind[0]
> ```

Oh I see.

mahdiqezlou commented:
I ran the code on TNG agian, after these modifications and job submission as below :

#!/bin/bash

#SBATCH --partition=short
#SBATCH --nodes=1
#SBATCH --tasks-per-node=8
#SBATCH --mem-per-cpu=3G
#SBATCH --cpus-per-task=4
#SBATCH -p short
#SBATCH --job-name="parallel"

#SBATCH --time=00:30:00

module load mpich
export OMP_NUM_THREADS=4
mpirun --map-by core python parallel.py

As you see in below, ranks did their jobs one after the other. As you said, it is suspicious that all ranks are done by a single CPU. It took 11 minutes which is still too long for 5 Line of sights and thresh = 0. Serial version took almost 30 minutes for 1000 sight lines and thresh = 0.

output  :

srun: Warning: can't honor --ntasks-per-node set to 8 which doesn't match the requested tasks
Rank at the beging : 0
Reloading from snapshot (will save to:  /rhome/mqezl001/bigdata/TNG/TNG100-1/postprocessing/ra
5  sightlines. resolution:  1.0  z= 1.9040895435327672
Rank =  0 is done !
Rank at the beging : 1
Reloading from snapshot (will save to:  /rhome/mqezl001/bigdata/TNG/TNG100-1/postprocessing/ra
5  sightlines. resolution:  1.0  z= 1.9040895435327672
Rank =  1 is done !
Rank at the beging : 2
Reloading from snapshot (will save to:  /rhome/mqezl001/bigdata/TNG/TNG100-1/postprocessing/ra
5  sightlines. resolution:  1.0  z= 1.9040895435327672
Rank =  2 is done !
Rank at the beging : 6
Reloading from snapshot (will save to:  /rhome/mqezl001/bigdata/TNG/TNG100-1/postprocessing/ra
5  sightlines. resolution:  1.0  z= 1.9040895435327672
Rank =  6 is done !
Rank at the beging : 3
Reloading from snapshot (will save to:  /rhome/mqezl001/bigdata/TNG/TNG100-1/postprocessing/ra
5  sightlines. resolution:  1.0  z= 1.9040895435327672
Rank =  3 is done !
Rank at the beging : 7
Reloading from snapshot (will save to:  /rhome/mqezl001/bigdata/TNG/TNG100-1/postprocessing/ra
5  sightlines. resolution:  1.0  z= 1.9040895435327672
Rank =  7 is done !
Rank at the beging : 4
Reloading from snapshot (will save to:  /rhome/mqezl001/bigdata/TNG/TNG100-1/postprocessing/ra
5  sightlines. resolution:  1.0  z= 1.9040895435327672
Rank =  4 is done !
Rank at the beging : 5
Reloading from snapshot (will save to:  /rhome/mqezl001/bigdata/TNG/TNG100-1/postprocessing/ra
5  sightlines. resolution:  1.0  z= 1.9040895435327672
Rank =  5 is done !

sbird commented:
hmm that is suspicious. At least the code is completing now, which is an improvement, but it seems like mpi4py is for some reason only actually running in serial. One thing you could do is to use the "--report-bindings" option to mpirun to report on which processors the code is running.

Another is to be yet more explicit to MPI what you want it to do. So

mpirun --report-bindings --map-by ppr:4:socket:pe=4 python parallel.py

where the --map-by ppr:4:socket:pe=4 means "put one rank on each socket and assign each rank a different set of 4 processors". (ppr = processors per rank)

mahdiqezlou commented:
> hmm that is suspicious. At least the code is completing now, which is an improvement, but it seems like mpi4py is for some reason only actually running in serial. One thing you could do is to use the "--report-bindings" option to mpirun to report on which processors the code is running.
> 
> Another is to be yet more explicit to MPI what you want it to do. So
> 
> mpirun --report-bindings --map-by ppr:4:socket:pe=4 python parallel.py
> 
> where the --map-by ppr:4:socket:pe=4 means "put one rank on each socket and assign each rank a different set of 4 processors". (ppr = processors per rank)

Thanks, I tried these options.
--report-bindings option is not recognized on cluster. By a brief search, I found this option is mostly used with openmpi rather mpich which I'm using. I have loaded openmpi and it is fine with that option, but I get other errors and from the printings I see it runs whole code just on rank=0.

I also added few comm.Barrier() before each broad casting and Reduce().  The  ranks are still being done one after each other with noticeable time separations. It seems they do not care about barrier!

sbird commented:
> --report-bindings option is not recognized on cluster. By a brief search, I found this option is mostly used with openmpi rather mpich which I'm using. I have loaded openmpi and it is fine with that option, but I get other errors and from the printings I see it runs whole code just on rank=0.

Ah well. mpich it is then.

> I also added few comm.Barrier() before each broad casting and Reduce(). The ranks are still being done one after each other with noticeable time separations. It seems they do not care about barrier!

That is extremely strange! Perhaps it is only the printing that is being serialised? Could you try adding a timestamp to your print statement to check that? If barrier() were not working MPI would be very badly broken.

mahdiqezlou commented:
> > --report-bindings option is not recognized on cluster. By a brief search, I found this option is mostly used with openmpi rather mpich which I'm using. I have loaded openmpi and it is fine with that option, but I get other errors and from the printings I see it runs whole code just on rank=0.
> 
> Ah well. mpich it is then.
> 
> > I also added few comm.Barrier() before each broad casting and Reduce(). The ranks are still being done one after each other with noticeable time separations. It seems they do not care about barrier!
> 
> That is extremely strange! Perhaps it is only the printing that is being serialised? Could you try adding a timestamp to your print statement to check that? If barrier() were not working MPI would be very badly broken.

After using timestamps, I found all ranks are run almost simultaneously and they are printed in the order they have finished their job. So It seems working for 5 sight-lines with threshold = 0
Now, I am going to run for thresh > 0 to see what happens.

mahdiqezlou commented:
It still has problems with thresh > 0, I got the errors like below after self.comm.Reduce() :

MPIDI_CH3U_Receive_data_found(129): Message from rank 4 and tag 11 truncated; 40 bytes received but buffer size is 20

Maybe the changes we made in " ind " have not been enough. I will take a more detailed look tomorrow morning.

mahdiqezlou commented:
> It still has problems with thresh > 0, I got the errors like below after self.comm.Reduce() :
> 
> File "/rhome/mqezl001/.local/lib/python3.6/site-packages/fake_spectra/spectra.py", line 665, in replace_not_DLA
> self.comm.Reduce(cdsum, col_den_added, op=self.MPI.SUM, root=0)
> File "MPI/Comm.pyx", line 705, in mpi4py.MPI.Comm.Reduce (src/mpi4py.MPI.c:99452)
> mpi4py.MPI.Exception: Message truncated, error stack:
> 
> PMPI_Reduce(1270).................: MPI_Reduce(sbuf=0x556a25d4dec0, rbuf=0x556a255cf300, count=5, MPI_FLOAT, MPI_SUM, root=0, MPI_COMM_WORLD) failed
> MPIR_Reduce_impl(1087)............:
> MPIR_Reduce_intra(848)............:
> MPIR_Reduce_impl(1087)............:
> MPIR_Reduce_intra(895)............:
> MPIR_Reduce_binomial(144).........:
> MPIDI_CH3U_Receive_data_found(129): Message from rank 4 and tag 11 truncated; 40 bytes received but buffer size is 20
> MPIR_Reduce_binomial(144).........:
> MPIDI_CH3U_Receive_data_found(129): Message from rank 1 and tag 11 truncated; 40 bytes received but buffer size is 20
> 
> Maybe the changes we made in " ind " have not been enough. I will take a more detailed look tomorrow morning.

In previous runs, I've set ndla=numlos=5. But after changing to ndla = 5 and numlos = 10, I do not get the error above and everything seems fine!
I am going to run for more numlos and ndla and higher threshold to see what happens.

sbird commented:
My guess is that it's having issues when ind is empty?

mahdiqezlou commented:
> My guess is that it's having issues when ind is empty?

Yes I see. So, I will increase threshold to higher values to see if the error happens again.

mahdiqezlou commented:
Humm ... When I check the result, I see the column densities are not actually larger than threshold( ~ 10**7) and also none of the ranks enter the while loop in replace_not_DLA(). So, as you said something is wrong with ind, though it's not empty here (size(ind) = ndla )

mahdiqezlou commented:
> Humm ... When I check the result, I see the column densities are not actually larger than threshold( ~ 10**7) and also none of the ranks enter the while loop in replace_not_DLA(). So, as you said something is wrong with ind, though it's not empty here (size(ind) = ndla )

I have put more print outs. As I mentioned above, column densities for each rank (right after col_den = self.compute_spectra() ) seem to be OK. But when they are added together after comm.Reduce, the result becomes crazy. I will have the first  ndla of components of col_den_added very large (~ 10**32) and consequently, as I have noticed in the previous comment, they cannot enter while loop (they already found the needed DLAs).
So, I just defined some test array (code pasted below) in the same place and used Reduce to add them together.  It works very well ! 
The number of wanted DLAs is ~ 10, so I cannot say the problem is with the size of col_den_added. If really appreciate any guess or suggestions!

        found = 0
        wanted = ndla
        size_ind = None
        cofm_DLA = np.empty_like(self.cofm)[:ndla, :]
        #Filter
        #Note: line does nothing
        col_den = self.compute_spectra(elem,ion,1215,False)
        col_den = np.array(col_den)
        cdsum=np.sum(col_den, axis=1)
        print('Rank =', self.rank, 'cdsum = ', cdsum)
        # A variable for comm.Reduce()
        col_den_added = np.zeros_like(cdsum, dtype='d')
        #ind = self.filter_DLA(col_den, thresh)
        
        ### Call manager rank here
        self.comm.Barrier()
        #print('rank = ', self.rank,'found = ', found, flush=True)
        self.comm.Reduce(cdsum, col_den_added, op=self.MPI.SUM, root=0)
        self.comm.Barrier()

        test = [1.0,1.0,1.0]
        test = np.array(test)
        test_added = np.zeros_like(test, dtype='d')
        print('Rank = ', self.rank, 'test =', test, flush = True)
        self.comm.Barrier()
        self.comm.Reduce(test, test_added, op=self.MPI.SUM, root = 0)
        self.comm.Barrier()

    
        #self.comm.Barrier()
        if self.rank == 0 :
            
            print('test_added = ', test_added, flush=True)

sbird commented on fake_spectra/spectra.py:
```diff
@@ -606,11 +613,30 @@ def replace_not_DLA(self, ndla, thresh=10**20.3, elem="H", ion=1):
-        ind = self.filter_DLA(col_den, thresh)
+        cdsum=np.sum(col_den, axis=1)
+        # A variable for comm.Reduce()
+        cdsum_added = np.zeros_like(cdsum, dtype='d')
```
Delete the dtype. If you use zeros_like it will automatically have the same type as the parent.

mahdiqezlou replied:
Yes, you are right. I have deleted at some point, the issue above has not been solved. However, you are right , that is redundant and I will delete that.

sbird replied:
Yes, I agree that it won't solve this issue. But if you were to change the type of cdsum in future it could cause a new bug.

mahdiqezlou commented:
> That is extremely weird. My only suggestion is that you use np.ravel() and np.ascontiguousarray()
> (https://docs.scipy.org/doc/numpy/reference/generated/numpy.ascontiguousarray.html)
> to ensure that the thing sent to Reduce is a flat 1D array contiguous in memory. Then reshape the thing you get out to have the same shape as it had before.

Sure! I am going to do that.

mahdiqezlou commented:
> > That is extremely weird. My only suggestion is that you use np.ravel() and np.ascontiguousarray()
> > (https://docs.scipy.org/doc/numpy/reference/generated/numpy.ascontiguousarray.html)
> > to ensure that the thing sent to Reduce is a flat 1D array contiguous in memory. Then reshape the thing you get out to have the same shape as it had before.
> 
> Sure! I am going to do that.

Wow!!! It seems working!! cdsum_added is no longer huge, It is reasonable!

sbird commented:
Aha!! That was a long shot so I'm glad it worked!

mahdiqezlou commented:
Hi Simeon, @sbird 

Could you let me know what you think about this version of the MPI feature ?

- The `parallel.py`  script is more or less the one I submit on clusters to run the MPI feature. 
- For compatibility, by default, the MPI feature is off unless some one passes MPI argument to `Spectra` or `RandSpectra` . 
- The way it distributes the work, as I have told before, is by distributing the hdf5 snapshots among ranks,. So, it only works for Illustris simulations.  The main reason I did so was the limited memory we have on a single a rank. If we want to distribute data among ranks by communications, we have to load all the snapshots on each rank and requires huge memory on each node. If you have a better idea, please let me know. It would be nice if we could use it for MP-Gadget outputs as well.
- As it it can be seen in the example script `parallel.py`  each rank saves the result on `savefile` file and the user just needs a simple script to add the column densities or optical depths. I have example scripts how one can do it. Or I can write a function in `spectra.py` to do the adding and then destroys the intermediate files.

sbird commented on fake_spectra/spectra.py:
```diff
@@ -310,10 +319,13 @@ def _save_multihash(self, save_array, grp):
                 del gg[str(key[-1])]
             except KeyError:
                 pass
+            """
```
I don't understand the need for commenting this?

mahdiqezlou replied:
I added this long time ago. But I guess it was because this line was deleting  the `ion=-1`  results.  I should check it again.

sbird replied:
That's strange: this line isn't doing anything with ion

mahdiqezlou replied:
I guess so. I reverted this.

sbird commented on fake_spectra/spectra.py:
```diff
@@ -673,28 +685,71 @@ def replace_not_DLA(self, ndla, thresh=10**20.3, elem="H", ion=1):
+                if self.rank == 0 :
+                    ind = self.filter_DLA(cdsum_added, thresh)
+                    size_ind = np.size(ind[0])
+                size_ind = self.comm.bcast(size_ind, root=0)
```
bcast vs Bcast?

mahdiqezlou replied:
To broadcast a numpy array, I should know the size of the array on all ranks.  So, first I broadcast the size of the array with `bcast` to all ranks and then the the array itself with `Bcast`.

sbird replied:
I see: bcast is for python arrays, Bcast for numpy arrays and mpi4py doesn't work the size out for itself. Thanks.

mahdiqezlou replied:
Exactly!

sbird commented on fake_spectra/parallel.py:
```diff
@@ -0,0 +1,47 @@
+import time
+import argparse
+
+def get_spec(num, base, savefile, savedir, res, numlos, thresh):
```
Docstring

mahdiqezlou replied:
It is much easier now to run the MPI feature, I added instructions on `README.md`. So I deleted this file.

sbird commented:
The code looks good, but could you also add the example script for combining the files? 

For the bigfile you actually can do something like:
            parts_per_rank = int(num_parts/size)
            bf = bigfile.BigFile(filename, 'r')
            bf["0/Density"][rank * parts_per_rank : (rank +1) * parts_per_rank]

and it will only use memory for the particles you actually read. Obviously the above needs some edge cases and so on!

mahdiqezlou commented:
> The code looks good, but could you also add the example script for combining the files?
> 
> For the bigfile you actually can do something like:
> parts_per_rank = int(num_parts/size)
> bf = bigfile.BigFile(filename, 'r')
> bf["0/Density"][rank * parts_per_rank : (rank +1) * parts_per_rank]
> 
> and it will only use memory for the particles you actually read. Obviously the above needs some edge cases and so on!

Oh, that's nice and maybe easier than hdf5 files then. I'll work on it.

mahdiqezlou commented:
Just checking something with you. Do you think it is safe to just modify this function so that returns the number of ranks ?
It seems to me the current selection for number of segments is not very special and could be changed, right ?

https://github.com/mahdiqezlou/fake_spectra-1/blob/bbc3935f4642a07909ca53dd6bb274e2384d2b07/fake_spectra/abstractsnapshot.py#L327

sbird commented:
I think so! I think the number of segments is just something I made up more or less at random for bigfile. Most things should work.

mahdiqezlou commented:
Sorry, It was a bug. It is solved now. I'll clean up and push it.

mahdiqezlou commented:
Here is the same sample of spectra showing the result is exactly the same as the non-MPI code (comparing overlapping solid and dashed lines) :

![spectra_bigfile_MPI](https://user-images.githubusercontent.com/44325095/117735739-eb968900-b1aa-11eb-82db-194bcc2aac58.png)

This test was run on 60 cMpc/h Full Hydro runs of MP-Gadget with pixel resolution of 6 km/s . 

The outputs are the spectra from each rank for the full spectra, they should be added. There are 2 options :

1. I copy a sample here and everyone can modify it for their own outputs.
2.  Using [MPI feature of hdf5](https://docs.h5py.org/en/stable/mpi.html) to do this immediately after each rank is done with it's work. 
        - It is very straightforward for me to implement it and I prefer this.
        - But requires people to compile their HDF5 with MPI feature on. I just checked the HDF5 on one of the Carnegie's cluster and found out by default it is not on. 
 
        Which one do you think is better ?

mahdiqezlou commented:
Something irrelevant : I should have said that the MPI feature is off as long as no MPI is passed to `Spectra` .

sbird commented:
Can you just do an MPI_Allreduce(MPI_SUM) equivalent at the end of the computation and then save from one rank? The MPI feature of HDF5 is very unstable on a lot of machines. You cannot rely on HDF5 being compiled correctly on most clusters (even Frontera has had some issues). This is why Gadget does not use HDF5's MPI feature, I think.

mahdiqezlou commented:
I see. Sure, I can use `MPI_SUM` but in terms of memory it will be an issue if someone wants to make a large map of the gas (when using griddedspectra.py). I have encountered this for TNG300, basically it did not fit on one single node.

sbird commented:
As I understand the code at the moment, each rank has its own copy of the sightlines, but only a subset of the particles? So if you do an MPI_SUM on the computed sightline tau at the end, this should not make any difference to the maximum memory usage. Or do I misunderstand how this works?

mahdiqezlou commented:
That's the right description of what is being dine here. You're right it should not make difference in terms of memory. I just have some vague memory that I had some problems with that on TNG300. For sure, I'm going to do that again and see how it works both on `bigfile` and `hdf5` files.

mahdiqezlou commented:
Hi @sbird 

- I have added the Allreduce() and it works well for both `bigfile` and `hdf5`. 
   - `bigfile` was tested on MP-Gadget 60cMpc/h simulations and the result was compared with the serial code.
    - `hdf5` was tested on TNG300 and the result was compared with my old MPI branch. 
- Also, `replace_not_DLA()` is working well and the spectra output seems good . For TNG :

![spectra_MPI_DLA](https://user-images.githubusercontent.com/44325095/117784598-9719fa80-b1f8-11eb-80a6-00fe53ab480b.png)

- Also, I added some notes on `README.md` on how to use it. 

Is anything else needed for this ?

sbird commented on fake_spectra/abstractsnapshot.py:
```diff
@@ -302,10 +334,14 @@ def get_data(self, part_type, blockname, segment):
+        self.parts_rank = ((self.get_npart()[part_type]//self.size)*np.ones(shape=(self.size,))).astype(int)
+        remainder = int(self.get_npart()[part_type]%self.size)
+        if remainder !=0 :
+            self.parts_rank[0:remainder] += 1
```
I don't totally understand this line: could you add a comment?

sbird replied:
Also chunck -> chunk is the right spelling

mahdiqezlou replied:
It is distributing particles among ranks, also divides  the particles on each rank into segments of  2*256**3 . I adopted this segmentation scheme from the previous version, I think it was for less memory  usage, right ?

Are the comments sufficient now ?

sbird replied:
I see - I was worried about the += 1 and the possibility of some edge cases losing a single particle, but I see now that this is fine because remainder < self.size. Looks good to me now!

sbird commented on fake_spectra/abstractsnapshot.py:
```diff
@@ -325,12 +361,13 @@ def _segment_to_partlist(self, part_type, segment):
-        return (one_segment*segment, None)
+        n_segments = self.get_n_segments(part_type)
+        one_segment = int(self.parts_rank[self.rank]/n_segments)
+        if self.rank ==0 :
```
Comment here would be useful too: it isn't immediately obvious although I figured it out.

sbird commented on fake_spectra/spectra.py:
```diff
@@ -760,20 +773,26 @@ def compute_spectra(self, elem, ion, ll, get_tau):
+        nsegments = self.snapshot_set.get_n_segments(part_type=0)
         arepo = (self.kernel_int == 2)
+        if arepo :
+           nsegments=1
```
I understand this as meaning that with the Voronoi mesh reconstruction there is no MPI splitting? Because the meshes mean that particles are not independent. Could you document this somewhere in Readme if so?

sbird commented on fake_spectra/spectra.py:
```diff
@@ -760,20 +773,26 @@ def compute_spectra(self, elem, ion, ll, get_tau):
+           # Add the spectra over all ranks
+           result = np.ascontiguousarray(result, np.float32)
+           result_all_ranks = np.zeros_like(result, dtype=np.float32)
+           self.comm.Allreduce(result, result_all_ranks, op=self.MPI.SUM)
```
Can you use MPI_IN_PLACE with the python Allreduce to save memory?

sbird commented on fake_spectra/spectra.py:
```diff
@@ -229,27 +240,29 @@ def __init__(self,num, base,cofm, axis, res=None, cdir=None, savefile="spectra.h
         """
         Saves spectra to a file, because they are slow to generate.
-        File is by default to be $snap_dir/snapdir_$snapnum/spectra.hdf5.
+        File is by default to be $snap_dir/snapdir_$snapnum/spectra_rank.hdf5.
```
This comment I think is out of date? It is saved to spectra.hdf5 but only by rank 0.

sbird commented on fake_spectra/spectra.py:
```diff
@@ -79,11 +79,22 @@ class Spectra:
+            self.comm = MPI.COMM_WORLD
+            self.rank = self.comm.Get_rank()
+            self.size = self.comm.Get_size()
+        else :
```
Need to add self.MPI = None here

sbird commented:
Works for me! Merging, thank you!
